batchflow/models/torch/base_mixins.py

```diff
@@ -221,6 +222,197 @@ class OptimalBatchSizeMixin:
+
+        batch_size = start_batch_size if isinstance(start_batch_size, list) else [start_batch_size]
+        notifier = Notifier(n_iters=None, bar=pbar,
+                            monitors=[{'source': batch_size, 'name': 'batch_size'}])
+        while True:
+            try:
+                notifier.update()
+                input = inputs or self.make_placeholder_data(batch_size[-1], to_device=False)
+                input = list(input) if isinstance(input, (tuple, list)) else [input]
+                input = [item[:batch_size[-1]] for item in input]
+
+                with GPUMemoryMonitor(frequency=frequency) as monitor:
+                    if method == 'train':
+                        target = targets or self.predict(inputs=input, outputs='predictions')
+                        target = list(target) if isinstance(target, (tuple, list)) else [target]
+                        target = [item[:batch_size[-1]] for item in target]
+
+                        _ = self.train(inputs=input, targets=target, microbatch_size=False)
+                    else:
+                        _ = self.predict(inputs=input, microbatch_size=False)
+
+                data = monitor.data
+                consumed_memory = np.mean(np.sort(data, axis=0)[-tail_size:])
+
+                # exit if max_memory was exceeded
+                if consumed_memory > max_memory:
+                    batch_size.append(batch_size[-1] // factor)
+                    break
+
+                batch_size.append(batch_size[-1] * factor)
+            except RuntimeError as exception:
+                if self.is_oom_error(exception):
+                    batch_size.append(batch_size[-1] // factor)
+                    break
+                raise # some other error not memory related
+
+            self.garbage_collection_cuda()
+        self.garbage_collection_cuda()
+        return batch_size[-1], batch_size[-1] * factor
+
+
+    def _run_binary_scaling(self, inputs=None, targets=None, low=2, high=None,
+                            start_batch_size=4, max_iters=25, pbar='n',
+                            method='train', max_memory=100, frequency=0.01,
+                            tail_size=20):
+        count = 0
+
+        # if None => make equal distance between low, start_batch_size and high
+        high = high if high is not None else 2 * start_batch_size - low
+
+        low = low if isinstance(low, list) else [low]
+        high = high if isinstance(high, list) else [high]
+        batch_size = start_batch_size if isinstance(start_batch_size, list) else [start_batch_size]
+
+        n_iters = int(np.ceil(np.log2(high[-1] - low[-1])))
+        notifier = Notifier(n_iters=n_iters, bar=pbar,
+                            monitors=[{'source': low, 'name': 'low'},
+                                      {'source': high, 'name': 'high'},
+                                      {'source': batch_size, 'name': 'batch_size'}])
```

Comment:
making all of the variables an instance of `list` so that Notifier can track them is not intuitive; while I like the idea, it should be commented in code